docs/developer/style_guide.md

```diff
@@ -0,0 +1,579 @@
+constant in the Constants module and this will make porting to other
+architectures (e.g. Cray) much simpler. Remember that all machine-specific code
+stays in the Utility routines. Similarly, when declaring a constant, use
+`1.234_dp` NOT `1.234d0`.
+
+
+### Free format
+
+Use free-format f90, with a maximum line length of 132 characters.
+
+
+### Spaces
+
+Use spaces between keywords e.g. `end if` not `endif`.
+
+
+### Code cleanly
+
+Obviously, to ensure correctness and make it easier for others to maintain the
+code. If the obvious way to code a routine is inefficient, whilst the efficient
+way is obscure, then this is allowed as long as the routine is well documented,
+both internally and in this Implementation Guide, with a discussion as to the
+design decisions made.
+
+
+### Be aware of vectorisation
+
+When coding, be aware that the code is often executed upon vector architectures.
+Therefore code with a style that enables good vectorisation - e.g. don't put
+`if` statements inside nested loops wherever possible.
+
+
+## Documentation
+
+
+### Internal
+
+Each module, based upon `ModuleTemplate.F90`, is to have a header block
+describing its overall purpose and function. Note that this header block also
+contains RCS directives to log changes to the code.
+
+Each routine, written to the template, is to have a header block describing
+its function, the variables required and how modified, the necessary conditions
+for correct operation, etc. These comments should be clear and easy to understand.
+
+Each routine is also to contain lots of comments within the code. Comments are
+to be either on the same line(s) as the feature being comment upon, or in a
+block of lines above the feature.
+
+Comment density should be comparable to code density.
+
+All comments are to be in English and obey standard rules of English grammar.
+
+
+### External
+
+A module (or any modifications) is only considered complete when additional
+documentation has been supplied for inclusion in the Implementation Guide. Each
+module is to be described by one chapter of the Implementation Guide, explaining
+the function of the module and each routine, the physics behind the code, the
+algorithms used, the key variables and data structures, and appropriate
+references to the literature.
+
+<!--  A documentation template is included in appendix  B. -->
+
+## Testing
+
+
+### Test Program
+
+Each module shall be tested for correctness before being added to the Bitbucket
+repository. To this end a set of tests shall be provided, which must compile and
+execute correctly in the CI system. The purpose of these tests is to verify the correct
+operation of every specified routine in the module.
+
+
+### Reference data
+
+The module test program shall come with both appropriate input datafiles and
+reference output results. Correct operation of the test shall be verified by
+comparing the actual output against the reference output results.
```

Comment:
References to `testcode.py` and auto-generation thereof?

Reply:
Yes, good idea.